lib/Runtime/Types/DictionaryPropertyDescriptor.h

```diff
@@ -6,62 +6,89 @@
 
 namespace Js
 {
+
+    namespace DictionaryPropertyDescriptorFlags
+    {
+        enum class Flags : unsigned char
+        {
+            None                         = 0,
+            PreventFalseReference        = 1 << 0,
+            IsShadowed                   = 1 << 1,
+            IsAccessor                   = 1 << 2,
+#if ENABLE_FIXED_FIELDS
+            IsInitialized                = 1 << 3,
+            IsOnlyOneAccessorInitialized = 1 << 4,
+            IsFixed                      = 1 << 5,
+            UsedAsFixed                  = 1 << 6,
+#endif
+        };
+
+        inline Flags operator& (Flags l, Flags r)
+        {
+            return static_cast<Flags>(static_cast<unsigned char>(l) & static_cast<unsigned char>(r));
+        }
+
+        inline Flags operator| (Flags l, Flags r)
+        {
+            return static_cast<Flags>(static_cast<unsigned char>(l) | static_cast<unsigned char>(r));
+        }
+
+        inline Flags operator~ (Flags flag)
+        {
+            return static_cast<Flags>(~static_cast<unsigned char>(flag));
+        }
+    }
+
     template <typename TPropertyIndex>
     class DictionaryPropertyDescriptor
     {
         template <typename T>
         friend class DictionaryPropertyDescriptor;
     public:
+
+        using Flags = DictionaryPropertyDescriptorFlags::Flags;
+
         DictionaryPropertyDescriptor(TPropertyIndex dataSlot, bool isInitialized = false, bool isFixed = false, bool usedAsFixed = false) :
-            Data(dataSlot), Getter(NoSlots), Setter(NoSlots), Attributes(PropertyDynamicTypeDefaults),
+            Data(dataSlot), Getter(NoSlots), Setter(NoSlots), Attributes(PropertyDynamicTypeDefaults), flags(Flags::PreventFalseReference
 #if ENABLE_FIXED_FIELDS
-            IsInitialized(isInitialized), IsOnlyOneAccessorInitialized(false), IsFixed(isFixed), UsedAsFixed(usedAsFixed),
+            | (isInitialized ? Flags::IsInitialized : Flags::None) | (isFixed ? Flags::IsFixed : Flags::None) | (usedAsFixed ? Flags::UsedAsFixed : Flags::None)
```

Comment:
a little hacky, but `(-isInitialized & Flags::IsInitialized) | (-isFixed & Flags::IsFixed) | (-usedAsFixed & Flags::UsedAsFixed)`
The idea is (I actually haven't tested) `-false & flag = 0 = Flags::None` `-true & flag = 0xFF & flag = flag`
At the very least, I would do something like
```cpp
inline Flags FlagIf(bool condition, Flags flag) { return condition ? flag : Flags::None; }
...
flags(Flags::PreventFalseReference |
#if ENABLE_FIXED_FIELDS
 | Flags::FlagIf(isInitialized, Flags::IsInitialized) | Flags::FlagIf(isFixed, Flags::IsFixed) | Flags::FlagIf(usedAsFixed, Flags::UsedAsFixed)
```
Then we can implement `FlagIf` however we like

Reply:
I'm not a fan of the using twos-complement bit fiddling here, I feel like that obfuscates more than it helps.

Regarding the `FlagIf`; I can't actually do `Flags::FlagIf` since the `enum class` isn't actually a class, so it can't have static functions. I could make it a static function on this class, but that strikes me as a little odd. If you think that it's clearer than using the ternary operator then I'm fine to do it though.

Reply:
I was able to do the FlagIf with
```cpp
 lib/Runtime/Types/DictionaryPropertyDescriptor.h | 12 ++++++++----
 1 file changed, 8 insertions(+), 4 deletions(-)

diff --git a/lib/Runtime/Types/DictionaryPropertyDescriptor.h b/lib/Runtime/Types/DictionaryPropertyDescriptor.h
index 4af444437..f7b59ae11 100644
--- a/lib/Runtime/Types/DictionaryPropertyDescriptor.h
+++ b/lib/Runtime/Types/DictionaryPropertyDescriptor.h
@@ -24,6 +24,10 @@ namespace Js
         };
 
         ENUM_CLASS_HELPERS(Flags, uint8);
+        inline Flags FlagIf(bool condition, Flags flag)
+        {
+            return condition ? flag : Flags::None;
+        }
     }
 
     template <typename TPropertyIndex>
@@ -38,28 +42,28 @@ namespace Js
         DictionaryPropertyDescriptor(TPropertyIndex dataSlot, bool isInitialized = false, bool isFixed = false, bool usedAsFixed = false) :
             Data(dataSlot), Getter(NoSlots), Setter(NoSlots), Attributes(PropertyDynamicTypeDefaults), flags(Flags::PreventFalseReference
 #if ENABLE_FIXED_FIELDS
-            | (isInitialized ? Flags::IsInitialized : Flags::None) | (isFixed ? Flags::IsFixed : Flags::None) | (usedAsFixed ? Flags::UsedAsFixed : Flags::None)
+            | FlagIf(isInitialized, Flags::IsInitialized) | FlagIf(isFixed, Flags::IsFixed) | FlagIf(usedAsFixed, Flags::UsedAsFixed)
 #endif
             ) { }
 
         DictionaryPropertyDescriptor(TPropertyIndex getterSlot, TPropertyIndex setterSlot, bool isInitialized = false, bool isFixed = false, bool usedAsFixed = false) :
             Data(NoSlots), Getter(getterSlot), Setter(setterSlot), Attributes(PropertyDynamicTypeDefaults), flags(Flags::PreventFalseReference | Flags::IsAccessor
 #if ENABLE_FIXED_FIELDS
-            | (isInitialized ? Flags::IsInitialized : Flags::None) | (isFixed ? Flags::IsFixed : Flags::None) | (usedAsFixed ? Flags::UsedAsFixed : Flags::None)
+            | FlagIf(isInitialized, Flags::IsInitialized) | FlagIf(isFixed, Flags::IsFixed) | FlagIf(usedAsFixed, Flags::UsedAsFixed)
 #endif
             ) { }
 
         DictionaryPropertyDescriptor(TPropertyIndex dataSlot, PropertyAttributes attributes, bool isInitialized = false, bool isFixed = false, bool usedAsFixed = false) :
             Data(dataSlot), Getter(NoSlots), Setter(NoSlots), Attributes(attributes), flags(Flags::PreventFalseReference
 #if ENABLE_FIXED_FIELDS
-            | (isInitialized ? Flags::IsInitialized : Flags::None) | (isFixed ? Flags::IsFixed : Flags::None) | (usedAsFixed ? Flags::UsedAsFixed : Flags::None)
+            | FlagIf(isInitialized, Flags::IsInitialized) | FlagIf(isFixed, Flags::IsFixed) | FlagIf(usedAsFixed, Flags::UsedAsFixed)
 #endif
             ) { }
 
         DictionaryPropertyDescriptor(TPropertyIndex getterSlot, TPropertyIndex setterSlot, PropertyAttributes attributes, bool isInitialized = false, bool isFixed = false, bool usedAsFixed = false) :
             Data(NoSlots), Getter(getterSlot), Setter(setterSlot), Attributes(attributes), flags(Flags::PreventFalseReference | Flags::IsAccessor
 #if ENABLE_FIXED_FIELDS
-            | (isInitialized ? Flags::IsInitialized : Flags::None) | (isFixed ? Flags::IsFixed : Flags::None) | (usedAsFixed ? Flags::UsedAsFixed : Flags::None)
+            | FlagIf(isInitialized, Flags::IsInitialized) | FlagIf(isFixed, Flags::IsFixed) | FlagIf(usedAsFixed, Flags::UsedAsFixed)
 #endif
             ) { }
 
```

Reply:
Huh; you're right. I thought I would have to prefix it with the peer namespace that contains the function, but that doesn't seem to be the case.